Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -69,8 +69,60 @@
 
 #if HAVE(WEB_TRANSPORT)
 
-static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&& connectionToWebProcess, URL&& url, WebKit::WebPageProxyIdentifier&& pageID, WebCore::ClientOrigin&& clientOrigin, sec_trust_t trust, sec_protocol_verify_complete_t completion)
+static bool leafCertificateMatchesWebTransportHash(sec_trust_t trust, const Vector<WebCore::WebTransportHash>& hashes)
 {
+    // https://www.w3.org/TR/webtransport/#verify-a-certificate-hash
+    RetainPtr secTrust = adoptCF(sec_trust_copy_ref(trust));
+    if (!secTrust)
+        return false;
+
+    RetainPtr chain = adoptCF(SecTrustCopyCertificateChain(secTrust.get()));
+    if (!chain || !CFArrayGetCount(chain.get()))
+        return false;
+
+    RetainPtr leafCertificate = checked_cf_cast<SecCertificateRef>(CFArrayGetValueAtIndex(chain.get(), 0));
+    if (!leafCertificate)
+        return false;
+
+    RetainPtr x509Data = adoptNS(bridge_cast(SecCertificateCopyData(leafCertificate.get())));
+    if (!x509Data)
+        return false;
+
+    std::array<uint8_t, CC_SHA256_DIGEST_LENGTH> sha2;
+    CC_SHA256([x509Data bytes], [x509Data length], sha2.data());
+
+    bool hashMatches = std::ranges::any_of(hashes, [&] (auto& hash) {
+        return hash.algorithm == "sha-256"_s && equalSpans(std::span(sha2), hash.value.span());
+    });
+    if (!hashMatches)
+        return false;
+
+    // https://www.w3.org/TR/webtransport/#custom-certificate-requirements
+    RetainPtr validityBegin = adoptCF(SecCertificateCopyNotValidBeforeDate(leafCertificate.get()));
+    if (!validityBegin)
+        return false;
+    RetainPtr validityEnd = adoptCF(SecCertificateCopyNotValidAfterDate(leafCertificate.get()));
+    if (!validityEnd)
+        return false;
+    CFAbsoluteTime currentTime = CFAbsoluteTimeGetCurrent();
+    CFAbsoluteTime beginTime = CFDateGetAbsoluteTime(validityBegin.get());
+    CFAbsoluteTime endTime = CFDateGetAbsoluteTime(validityEnd.get());
+    if (currentTime < beginTime || currentTime > endTime)
+        return false;
+
+    CFTimeInterval validityPeriod = endTime - beginTime;
+    constexpr double twoWeeks = 2 * 7 * 24 * 60 * 60;
```

Comment:
nit: could use something like `2 * 7 * secondsPerDay` (but it's only marginally more readable)